Fix #208: Specify numeric and string constraints

lgolding opened the pull request:
We reviewed the string length and numeric value constraints on all SARIF properties, as well as the length constraints on all array properties, and the constraints on the number of sub-properties in properties such as `run.files`.

The following arrays were required to be non-empty if present, and I did not change them:
- `sarifLog.runs`
- `result.locations`
- `result.stacks`
- `result.relatedLocations`
- `codeFlow.locations`
- `stack.Frames`

`result.CodeFlows` was not required to be non-empty, so I changed it for consistency with `result.stacks`.

The following arrays are allowed to be empty:
- `run.results` (a special case)
- `result.fixes` (why?)
- `exception.innerException`

As the text now stands, empty objects (`{}`) are optional, but most empty arrays (`[]`) are forbidden. This doesn't make sense to me. Let's discuss.

Also:
- Adjusted several places where we expressed the same thing using slightly different phrasing or punctuation in different places. In all such cases, I unified on the construct that was used most often.

Commit:
Fix #208: Specify numeric and string constraints

We reviewed the string length and numeric value constraints on all SARIF
properties, as well as the length constraints on all array properties,
and the constraints on the number of sub-properties in properties
such as `run.files`.

The following arrays were required to be non-empty if present, and
I did not change them:

* `sarifLog.runs`
* `result.locations`
* `result.stacks`
* `result.relatedLocations`
* `codeFlow.locations`
* `stack.Frames`

`result.CodeFlows` was not required to be non-empty, so I changed it for
consistency with `result.stacks`.

The following arrays that are allowed to be empty:
* `run.results` (a special case)
* `result.fixes` (why?)
* `exception.innerException`

As the text now stands, empty objects ( {} ) are optional, but most
empty arrays are forbidden. This doesn't make sense to me. Let's discuss.

Also:
* Adjusted several places where we expressed the same thing using
  slightly different words in different places. In all such cases,
  I unified on the construct that was used most often.

Commit:
Merge remote-tracking branch 'upstream/master'

michaelcfanning commented:
@lgolding reviewing. result.fixes and exception.innerException look like bugs (assuming we don't change the 'empty arrays forbidden' standard.

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -155,14 +155,41 @@ property, for example, "this is a source file".
+
+Certain properties in this specification are defined to be JSON arrays.
+Examples are the `run.toolNotifications` property (&sect;[#run-toolNotifications]) and the
+`file.hash` property (&sect;[#file-hashes]).
```
> hash [](start = 6, length = 4)

s/be 'file.hashes' property #Resolved

lgolding replied:
fixed.

---
In reply to: [75180701](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75180701) [](ancestors = 75180701)

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -758,7 +786,7 @@ by means of the `result`'s `ruleId` property (&sect;[#result-ruleId]) or `ruleKe
 ### `properties` property { #run-properties }
 
-A `run` object may contain a property named `properties`, whose value is property bag.
+A `run` object may contain a property named `properties` whose value is property bag.
```
> is property bag. [](start = 69, length = 16)

s/be ' is a property bag' #Resolved

lgolding replied:
fixed

---
In reply to: [75180912](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75180912) [](ancestors = 75180912)

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -898,7 +926,7 @@ the precise version number of the logging component that produced the invalid ou
 ### `properties` property { #tool-properties }
 
-A `tool` object may contain a property named `properties`, whose value is property bag.
+A `tool` object may contain a property named `properties` whose value is property bag.
```
> is property bag [](start = 70, length = 15)

again 'is a property bag'. search for this pattern, i won't repeat the comment if i see it again. :) #Resolved

lgolding replied:
Wow, there were half a dozen of those. Good catch. All fixed (and added section reference to `&sect;[#property-bags]`)

---
In reply to: [75181111](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75181111) [](ancestors = 75181111)

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -913,16 +941,28 @@ An `invocation` object may contain a property named `commandLine` whose value is
 containing the runtime parameters with which the tool was invoked.
```
> was invoked [](start = 54, length = 11)

need to clearly specify whether the .exe or script name is provided as a first argument. i tend to think this should be the case. the text here suggests only the parameters are provided. #Resolved

lgolding replied:
Ok, let's go with that it "shall" include the exe/script name, but with the caveat that path information can be redacted (as it is in the example).

---
In reply to: [75181386](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75181386) [](ancestors = 75181386)

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -913,16 +941,28 @@ An `invocation` object may contain a property named `commandLine` whose value is
 and to verify that the results reported in the log file were generated by an appropriate invocation of the tool.
 ~
 
+The value of the `commandLine` property may be an empty string.
```
> property may be an empty string [](start = 31, length = 31)

if my comment above stands, it should not be empty, it would either be absent or contain the exe/script name. #Resolved

lgolding replied:
Not if the tool is invoked in some way other than by a command line. I'll clarify.

---
In reply to: [75181477](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75181477) [](ancestors = 75181477)

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -1029,9 +1065,11 @@ Absolute path names can reveal information that might be sensitive.
+The property names in this object shall contain the names of all the environment variables in the tool's
+execution environment. The value of each property shall be a string containing the value of the specified
 environment variable.
+If the value of the environment variable is an empty string,
```
> environment variable is an empty string [](start = 20, length = 39)

In Windows, it is not possible to define an environment variable with an empty string. Doing so removes the variable from the environment. #Resolved

lgolding replied:
TIL. And on Unix, an environment variable _can_ have an empty value. I added a NOTE to that effect.

---
In reply to: [75182122](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75182122) [](ancestors = 75182122)

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -1164,7 +1201,7 @@ necessarily contain enough information to do so.
 
 Depending on the circumstances, a `file` object either shall, may, or shall not contain a property named `offset`
-whose value is an integer.
+whose value is a non-negative integer.
```
> non-negative [](start = 17, length = 12)

we need a schema change for this, are we tracking those anywhere? #Resolved

lgolding replied:
Yes, we have Microsoft/sarif-sdk#484, "Update schema per constraints review".

---
In reply to: [75182354](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75182354) [](ancestors = 75182354)

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -1656,8 +1699,9 @@ or include them in bug counts.
-the developer has suppressed it in the source code (see &sect;[#result-suppressionStates-suppressedInSource]).
+In this version of the SARIF standard, the only supported reasons for suppressing a result is that
+the developer has suppressed it in the source code (see &sect;[#result-suppressionStates-suppressedInSource])
+or that it is marked as "suppressed" an external database (see &sect;[#result-suppressionStates-suppressedExternally]).
```
> suppressed [](start = 25, length = 10)

why do you enquote here and not elsewhere in the para? probably can remove the quotes. might be best to use 'external persisted store' or something more general than database. we could be working with a simple baseline written to a text file, for example. we could use language lile 'external store, such as a database' perhaps. #Resolved

lgolding replied:
fixed, fixed.

---
In reply to: [75183149](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75183149) [](ancestors = 75183149)

michaelcfanning commented:
:clock1:

lgolding commented:
The spec does not prohibit empty arrays. It does say how an array must be treated if it is empty:

"Unless otherwise specified in the description of a specific property, if any such array is empty, then either the property may be represented as an empty array `[]`, or it may be absent."

Every array-valued property specifies whether it is allowed to be empty, for example:

"A `run` object may contain a property named `configurationNotifications` whose value is an array of zero or more `notification` objects."

However:

* `result.fixes` did _not_ specify whether it could be empty. I changed it to "one or more".
* `exception.innerExceptions` did _not_ specify whether it could be empty. I changed it to "one or more".

---
In reply to: [240505136](https://github.com/sarif-standard/sarif-spec/pull/256#issuecomment-240505136) [](ancestors = 240505136)

Commit:
PR feedback on #208

Commit:
Claryify that commandLine is only for command line tools

michaelcfanning commented on document/FileFormat.mdk:
```diff
@@ -910,19 +938,25 @@ An `invocation` object contains information describing the invocation of the ana
-containing the runtime parameters with which the tool was invoked.
-For command line tools, this string may consist of the completely specified command line used to invoke the tool.
+containing the completely specified command line used to invoke the tool,
+starting with the path to the tool's executable or script file.
```
> starting with the path [](start = 0, length = 22)

The command-line doesn't need to start with the path, the user could simply invoke the tool or script name (which could result in the %PATH% var being consulted). Note that this is why we provide the filePath member, so that the log file can specify the complete path to the tool, if the path is resolved at runtime. The commandLine should precisely match the actual commandLine used to invoke the tool.

lgolding replied:
Good catch. Changed to: "... starting with the name of the tool's executable or script file, optionally qualified by the relative or absolute path to the file."

---
In reply to: [75796568](https://github.com/sarif-standard/sarif-spec/pull/256#discussion_r75796568) [](ancestors = 75796568)

michaelcfanning commented:
> such as the `System.AggregateException` class from the .NET Framework.

+1 Nice change. Modulo correction around specifying path to tool in commandLine...

---
Refers to: document/FileFormat.mdk:3322 in 239514c. [](commit_id = 239514ceab40a12cdcd62be233890fadbb851395, deletion_comment = False)

Commit:
Clarify that command line need not start with a path.
